docs/src/rfc/text/0017-aml-parser.md

```diff
@@ -0,0 +1,282 @@
+with less support for actually walking through and modifying the firmware-side AML object tree.
+
+There is ongoing conversation with the owner of the `acpi` crate about
+borrowing certain implementations and modifying the public interfaces
+to be more friendly to the Patina ACPI implementation.
+This conversation can be tracked through a [Github issue in the `acpi` crate](https://github.com/rust-osdev/acpi/issues/260).
+
+## Rust Code
+
+### AML Handles
+
+By spec definition, an AML handle is an opaque handle returned from opening a DSDT or SSDT,
+on which AML traversal and patching operations can be performed.
+
+Internally, the each `AmlHandle` object, aliased as `AmlSdtHandleInternal`,
+represents a cursor within an AML stream. Each handle can be conceptualized as a "node"
+in the AML object tree of parent-child-relationships.
+
+```rust
+pub(crate) struct AmlSdtHandleInternal {
+    table_key: TableKey,
+    offset: usize,
+    size: usize,
+    modified: bool,
+    byte_encoding: AmlByteEncoding,
+    parent_end: Option<usize>,
+}
+
+impl AmlSdtHandleInternal {
+    fn new(table_key: TableKey, offset: usize, size: usize) -> Self {
+        Self { table_key, offset, size, modified: false }
+    }
+}
+
+pub type AmlHandle = AmlSdtHandleInternal;
+```
+
+The `size` of an `AmlSdtHandleInternal` refers to its full size, including any children (`TermList`).
+
+The `offset` refers to its offset with the AML stream of the table.
+Offset 0 is the start of the AML stream, and the highest offset is at the end of `table_length`.
+
+`modified` ensures the corresponding table (which can be retrieved through `table_key`) has an updated checksum
+if the contents are modified by `set_option`.
+
+Each handle stores the `parent_end` of its parent node, which is the parent's `size` + `offset` (useful for retrieving siblings).
+
+Each handle also stores its own byte encoding, specifying its own layout:
+
+```rust
+#[derive(PartialEq, Eq, Hash)]
+struct AmlByteEncoding {
+    opcode: AmlOpcode,
+    operands: Vec<AmlOperand>,
+    attributes: AmlOpAttributes,
+}
+
+bitflags! {
+    pub struct AmlOpAttributes: u32 {
+        /// If opcode has a pkg_length field.
+        const HAS_PKG_LENGTH  = 0x0000_0001;
+
+        /// If opcode has children.
+        const HAS_CHILD_OBJ   = 0x0000_0002;
+    }
+}
+
+/// Represents the possible opcodes. 
+pub enum AmlOpcode {
+    BaseOp(BaseOpcode),
+    ExtOp(ExtOpcode),
+}
+
+pub enum BaseOpcode {
+    ZerOp,
+    AliasOp,
+    ...
+}
+
+pub enum ExtOpcode {
+    MutexOp,
+    EventOp,
+    ...
+}
+
+pub enum AmlOperand {
+    Opcode(AmlOpcode),
+    Name(AmlNameString), // Represents a NameString (AML path). Not to be confused with a string literal 
+    ...
+}
+```
+
+### AML Trait Interface
+
+The `AmlParser` service generally derives from the ACPI SDT protocol, and allows for traversal of the AML object tree.
+
+```rust
+pub(crate) trait AmlParser {  

```

Comment:
This feels, and as described, a direct derivation from the SDT protocol, which is obviously needed, though is a very C approach. For the Rust side, would it make sense to just return an iterator that does the get_child/get_sibling in the iteration itself? Feels like that would be more useful and safer. Then its debatable about whether you have an iter_mut that lets the caller directly change it or you force them to call set_option from the iter once they find what they want.

But, in general, trying to think about how we can ensure that this API is defined for a Rust interface and the C protocol is supported but not the primary interface driver.